### PR TITLE
Make management command to flag duplicate IDs

### PR DESCRIPTION
**Pull request by esoergel**
https://trello.com/c/Av4Y5ltP/264-duplicate-id-flagging
This PR pulls out the code which identifies duplicate IDs and uses that for a management command which flags all of those cases for later update.

@sravfeyn

### CODE REVIEW

File: custom/enikshay/duplicate_ids.py [hunk 1]
```diff
@@ -12,6 +14,68 @@
 from .const import VOUCHER_ID, ENIKSHAY_ID
 
 
+class DuplicateIdsResult(AggregationResult):
```
[the remark this answers is not in the document]
Comment (proteusvacuum) [reply custom/enikshay/duplicate_ids.py]: Thanks for trying! I'm interested in understanding how we can do stuff like this better in the future, so this is all good to know.

File: custom/enikshay/duplicate_ids.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from dimagi.utils.couch.database import iter_docs
+
+from .const import VOUCHER_ID, ENIKSHAY_ID
+
+
+def get_cases_with_duplicate_ids(domain, case_type, all_case_ids):
+    accessor = CaseAccessors(domain)
+    id_property = {'voucher': VOUCHER_ID, 'person': ENIKSHAY_ID}[case_type]
```
Comment (sravfeyn) [inline custom/enikshay/duplicate_ids.py]: Isn't below be more direct and readable?
```
id_property = VOUCER_ID if case_type == 'voucher' else ENIKSHAY_ID
```

File: custom/enikshay/duplicate_ids.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    return context
+
+
+def add_debug_info_to_cases(bad_cases, full_debug_info):
+    _add_form_info_to_cases(bad_cases, full_debug_info)
+    _add_user_info_to_cases(bad_cases)
+
+
+def _add_form_info_to_cases(bad_cases, full_debug_info):
+    for case in bad_cases[:None if full_debug_info else 300]:
```
Comment (sravfeyn) [inline custom/enikshay/duplicate_ids.py]: What is 300?

File: custom/enikshay/duplicate_ids.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    auth_usernames = {user_doc['_id']: user_doc['username'] for user_doc in
+                      iter_docs(CommCareUser.get_db(), auth_user_ids)}
+    for case in bad_cases:
+        user_dict = user_info.get(case.get('form_user_id'))
+        if user_dict:
+            case['username'] = user_dict['username']
+            device_id = case['form_device_id']
+            if device_id == 'Formplayer':
+                if case['form_user_id'] == case['auth_user_id']:
+                    device_id = "WebAppsLogin"
```
Comment (sravfeyn) [inline custom/enikshay/duplicate_ids.py]: Can this be declared as a constant somewhere?

File: custom/enikshay/duplicate_ids.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+from corehq.apps.users.models import CommCareUser
+from corehq.form_processor.backends.sql.dbaccessors import CaseAccessorSQL
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from dimagi.utils.couch.database import iter_docs
+
+from .const import VOUCHER_ID, ENIKSHAY_ID
+
+
+def get_cases_with_duplicate_ids(domain, case_type, all_case_ids):
```
Comment (sravfeyn) [inline custom/enikshay/duplicate_ids.py]: Doc comments on these functions will be useful for others.

**Review by sravfeyn — APPROVED** [review]
Looks good, very minor comments. Feel free to self-merge after addressing them

File: custom/enikshay/duplicate_ids.py (hunk 2 above)
Comment (snopoke) [reply custom/enikshay/duplicate_ids.py]: unless you want to raise an exception when the case type is neither `voucher` nor `person`

File: custom/enikshay/duplicate_ids.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def add_debug_info_to_cases(bad_cases, full_debug_info):
+    _add_form_info_to_cases(bad_cases, full_debug_info)
+    _add_user_info_to_cases(bad_cases)
+
+
+def _add_form_info_to_cases(bad_cases, full_debug_info):
+    for case in bad_cases[:None if full_debug_info else 300]:
+        form = CaseAccessorSQL.get_transactions(case['case_id'])[0].form
```
Comment (snopoke) [inline custom/enikshay/duplicate_ids.py]: If you're wanting to get the 'opening' form you could use `case.get_opening_transactions()[0]`

File: custom/enikshay/duplicate_ids.py (hunk 3 above)
Comment (esoergel) [reply custom/enikshay/duplicate_ids.py]: It's an arbitrary number of cases for which to include expensive debug info.  I moved this to the view and added a comment

File: custom/enikshay/duplicate_ids.py (hunk 4 above)
Comment (esoergel) [reply custom/enikshay/duplicate_ids.py]: I'm not sure that's appropriate.  This is a hack that I wouldn't want to grant that much legitimacy.  WebApps uses different device IDs in restores and in form submissions, and this code takes a guess as to what WebApps probably used in the restore.  I added a comment about it.

File: custom/enikshay/duplicate_ids.py (hunk 5 above)
Comment (esoergel) [reply custom/enikshay/duplicate_ids.py]: I just added a file-level explanation of the issue that these functions address.  With that context, my hope is that the function names can be descriptive enough.

File: custom/enikshay/duplicate_ids.py (hunk 6 above)
Comment (esoergel) [reply custom/enikshay/duplicate_ids.py]: these are case stubs, so unfortunately that's not available.  Would it be preferable to do something like this?
```python
CaseAccessorSQL.get_transactions_by_type(case_id, CaseTransaction.TYPE_CASE_CREATE)
```

Comment (snopoke) [reply custom/enikshay/duplicate_ids.py]: I think that is preferable since it will load a lot less transactions from the DB.

Comment (esoergel) [reply custom/enikshay/duplicate_ids.py]: Updated.  Thanks for the tip!

**Review by millerdev — APPROVED** [review]
Only reviewed last three commits, as requested by @esoergel

**Pull request merged**